### PR TITLE
Fix/handle nil licensed users in terminate

### PR DESCRIPTION
**Pull request by george42-ctds**
Fixing:
```
2024/03/18 16:49:33 Terminating workspace for user ABC
2024/03/18 16:49:33 Checking for gen3 license items for user: ABC
2024/03/18 16:49:33 Error in items for user query: InvalidParameter: 2 validation error(s) found.
- minimum field size of 3, QueryInput.IndexName.
- minimum field size of 3, QueryInput.TableName.
2024/03/18 16:49:33 InvalidParameter: 2 validation error(s) found.
- minimum field size of 3, QueryInput.IndexName.
- minimum field size of 3, QueryInput.TableName.
2024/03/18 16:49:33 Debug: Active gen3 license user maps <nil>
2024/03/18 16:49:33 http: panic serving 172.27.100.165:43980: runtime error: invalid memory address or nil pointer dereference
goroutine 3528 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1802 +0xb9
panic({0x18ee0a0, 0x2e004c0})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
github.com/uc-cdis/hatchery/hatchery.terminate({0x1f9cd50, 0xc000add6c0}, 0xc00009ec00)
	/go/src/github.com/uc-cdis/hatchery/hatchery/hatchery.go:499 +0x259
net/http.HandlerFunc.ServeHTTP(0xc0006e0870, {0x1f9cd50, 0xc000add6c0}, 0xc0007978a8)
	/usr/local/go/src/net/http/server.go:2047 +0x2f
net/http.(*ServeMux).ServeHTTP(0x1f9c2d0, {0x1f9cd50, 0xc000add6c0}, 0xc00009ec00)
	/usr/local/go/src/net/http/server.go:2425 +0x149
gopkg.in/DataDog/dd-trace-go.v1/contrib/internal/httputil.TraceAndServe({0x1f76da0, 0xc0005f72c0}, 0xc000797a40)
	/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.33.0/contrib/internal/httputil/trace.go:57 +0xa47
gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http.(*ServeMux).ServeHTTP(0xc0006d50a0, {0x1f9c2d0, 0xc0004fa0e0}, 0xc00009eb00)
	/go/pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.33.0/contrib/net/http/http.go:49 +0x17e
net/http.serverHandler.ServeHTTP({0xc00012b7a0}, {0x1f9c2d0, 0xc0004fa0e0}, 0xc00009eb00)
	/usr/local/go/src/net/http/server.go:2879 +0x43b
net/http.(*conn).serve(0xc0004846e0, {0x1fa4180, 0xc0006e4cf0})
	/usr/local/go/src/net/http/server.go:1930 +0xb08
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3034 +0x4e8
```

### Bug Fixes

* Allow `/terminate` to complete when dynamodb query returns nil values

### CODE REVIEW

File: hatchery/hatchery.go [hunk 1]
```diff
@@ -496,7 +496,7 @@ func terminate(w http.ResponseWriter, r *http.Request) {
 		Config.Logger.Printf(userlicerr.Error())
 	}
 	Config.Logger.Printf("Debug: Active gen3 license user maps %v", activeGen3LicenseUsers)
-	if len(*activeGen3LicenseUsers) == 0 {
+	if activeGen3LicenseUsers == nil || len(*activeGen3LicenseUsers) == 0 {
```
Comment (paulineribeyre) [inline hatchery/hatchery.go:499]: This would work, but I think the `getLicenseUserMapsForUser` function should return a pointer to an empty list instead of nil when the user doesn't have a license.
And the edge case should be added to the `Test_GetActiveGen3LicenseUserMaps` unit test

Comment (paulineribeyre) [inline hatchery/hatchery.go:499]: @george42-ctds said:
> I think the edge case is when /terminate tries to query the table but the table does not exist in the environment. That shows up as errors in the logs.
for that case, should the getActiveGen3LicenseUserMaps return an empty list instead of nil?

Ok, yes i do think `getActiveGen3LicenseUserMaps` should check if the config exists before trying to query, and in any case, return an empty list if the user doesn't have a license.

I think the `getActiveGen3LicenseUserMaps` function should return `gen3LicenseUserMaps []Gen3LicenseUserMap` instead of `gen3LicenseUserMaps *[]Gen3LicenseUserMap`. That should remove a lot of the confusion and the returned type will be enforced, so it will always return a list and the calling functions won't have to check the returned value.

Comment (george42-ctds) [reply hatchery/hatchery.go]: Good ideas! 

- output type is updated
- empty list is returned for any DB errors or for no active users
- test case added for function to return empty list (triggered by license not enabled in config).

Comment (george42-ctds) [reply hatchery/hatchery.go]: Additional check has been added.

File: hatchery/gen3licenseusermaps.go [hunk 2]
```diff
@@ -99,11 +99,16 @@ func getItemsFromQuery(dbconfig *DbConfig, queryInput *dynamodb.QueryInput) ([]m
...
 
-var getActiveGen3LicenseUserMaps = func(dbconfig *DbConfig, container Container) (gen3LicenseUserMaps *[]Gen3LicenseUserMap, err error) {
+var getActiveGen3LicenseUserMaps = func(dbconfig *DbConfig, container Container) (gen3LicenseUserMaps []Gen3LicenseUserMap, err error) {
 	// Query the table to get all active gen3 license user map items
+	emptyList := []Gen3LicenseUserMap{}
 
 	targetEnvironment := os.Getenv("GEN3_ENDPOINT")
-	validateContainerLicenseInfo(container.Name, container.License)
+	ok := validateContainerLicenseInfo(container.Name, container.License)
+	if !ok {
```
Comment (paulineribeyre) [inline hatchery/gen3licenseusermaps.go:108]: by convention, the way to do this in Go is with errors

this function should return an error instead of a boolean https://github.com/uc-cdis/hatchery/blob/b32a3bc/hatchery/gen3licenseusermaps.go#L49
and the calling functions can choose what to do with the error (throw it or not - here, ignore it and return an empty list)

and side question - should the returned value by checked here too? https://github.com/uc-cdis/hatchery/blob/b32a3bc/hatchery/gen3licenseusermaps.go#L301
(if you use an error instead of a bool, i think Go refuses to build if you don't check the returned error, which helps not to miss something like this)

File: hatchery/gen3licenseusermaps.go [hunk 3]
```diff
@@ -99,11 +99,20 @@ func getItemsFromQuery(dbconfig *DbConfig, queryInput *dynamodb.QueryInput) ([]m
...
+		Config.Logger.Printf("Gen3License table info for container is not configured.")
```
Comment (paulineribeyre) [inline hatchery/gen3licenseusermaps.go:109]:
```suggestion
		Config.Logger.Printf("Gen3License table info for container is not configured or misconfigured.")
```

File: hatchery/gen3licenseusermaps.go (hunk 2 above)
Comment (george42-ctds) [reply hatchery/gen3licenseusermaps.go]: Nice! Return type for `validateContainerLicenseInfo` has been updated to `error`.

**Review by paulineribeyre — DISMISSED** [review]
awesome, thanks @george42-ctds!

**Review by paulineribeyre — APPROVED** [review]
